{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bff91b09",
   "metadata": {},
   "source": [
    "# Exercises Week 10-11: Graded exercise session (part B)\n",
    "\n",
    "**Course**: [Life Sciences engineering: genome to function](https://go.epfl.ch/BIO-411) (BIO-411)\n",
    "\n",
    "**Professors**:  _Gönczy Pierre_, _Naef Felix_, _McCabe Brian Donal_\n",
    "\n",
    "SSV, MA, 2025-2026"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "867ce50e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.decomposition import PCA\n",
    "import plotly.express as px\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from scipy.integrate import odeint\n",
    "import matplotlib.pyplot as plt\n",
    "from ipywidgets import interact, fixed\n",
    "import scipy\n",
    "from scipy.stats import beta\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import math\n",
    "import seaborn as sns\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e04f778e",
   "metadata": {},
   "source": [
    "**Provide answers in the form of code, figures (if relevant) and short descriptions (in markdown cells) in those notebooks. Submit your notebook to Moodle, please make sure to execute every cell.  \n",
    "We very strongly advise each member of the group to work on the whole notebook on their own to understand it in details, as it comes out that in the exam this subject is often poorly understood.  \n",
    "The theory related to Exercise 2 Question 2 will be discussed in the lecture of 27th November.**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4dcb272",
   "metadata": {},
   "source": [
    "### Exercise 1: Simulation of cell cycle gene (RNA) expression"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecaa0b5a",
   "metadata": {},
   "source": [
    "In this exercise, we focus on the transcription and splicing of a pre-mRNA to produce mRNA. We describe this process using a two-dimensional Ordinary Differential Equation (ODE) with **time-dependent** transcription and degradation rates to model cell cycle dynamics.\n",
    "\n",
    "The system is defined by the following ODEs:\n",
    "\n",
    "\\begin{array}{ll}\n",
    "\\frac{dP}{dt} = s(t) - \\rho P \\\\  \n",
    "\\\\\n",
    "\\frac{dM}{dt} = \\rho P - k(t) M \\\\\n",
    "\\end{array}\n",
    "\n",
    "where  $P$ and $M$ represent the concentrations of pre-mRNA and mRNA, respectively. The frequency is defined by $\\omega=\\frac{2\\pi}{T}$, taking a cell cycle period $T=20h$.  \n",
    "\n",
    "$s(t)$ is the transcription rate, which is now taken either as constant $s(t)=s_{0}$ or as a periodic function of time:  \n",
    "\n",
    "\\begin{equation*}\n",
    "s(t) = s_0(1 + \\epsilon_s \\cos(\\omega t)) \n",
    "\\end{equation*}  \n",
    "with relative amplitude $0\\leq \\epsilon_s\\leq 1$. Note that the peak of $s(t)$ is at $t=0$.  \n",
    " \n",
    "Similarly, we will consider the degradation rate $k(t)$ either as constant $k(t)= k_0$ or a periodic function of time:  \n",
    "\\begin{equation*}\n",
    "k(t)=k_0(1 + \\epsilon_k \\cos(\\omega (t-t_k))\n",
    "\\end{equation*}  \n",
    "with $0\\leq\\epsilon_k\\leq 1$, and the maximum degradation rate at $t=t_k$.\n",
    "\n",
    "$\\rho$ represents the constant splicing rate of pre-mRNA $P$ into mRNA $M$."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebecf369",
   "metadata": {},
   "source": [
    "### Question 1\n",
    "1. Adapt the code from Week 9 exercises to simulate this system with periodic transcription and degradation rates. Modify the interactive widget such that you can vary the new parameters and plot $P(t)$ and $M(t)$.  \n",
    "\n",
    "2. Keep the degradation rate constant (set $\\epsilon_k$=0) and vary $k_0$ and $\\epsilon_s$.\n",
    "    * How does the phase delay between $M(t)$ and $P(t)$ changes with different values of $k_0$?\n",
    "    * How does the amplitude (use the ratio between max and min) of $M(t)$ changes with different values of $k_0$? \n",
    "    \n",
    "    Make sure you span a relevant range for $k_0^{-1}$ (*i.e.* from 10 minutes to several hours).  \n",
    "\n",
    "3. Now introduce periodic degradation by varying $\\epsilon_k$. \n",
    "    * Show that you can obtain a periodic mRNA accumulation when the synthesis is constant ($\\epsilon_s=0$). \n",
    "\n",
    "**Since the widget results are not kept upon saving the notebook, when showing plots is necessary for your answer, please take a screenshot of the obtained results and include it in your answer, together with the values of the parameters you used to get the behaviour you want to show.**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94467283",
   "metadata": {},
   "source": [
    "### Exercise 2: Cell cycle (post-)transcriptional regulation of gene expression in human RPE1 cells"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d4c6d92",
   "metadata": {},
   "source": [
    "In this exercise, we will analyze RNA-seq data from normally dividing human RPE1 cells (cell cycle period of $T=20h$).\n",
    " The provided data has two replicates and is binned into 50 equally spaced time points ($dt=0.4h=24min$). Similar to the Week 9 exercises on RNA-seq data, gene expression was quantified at the intron (i.e., unspliced) and exon (i.e., spliced) levels.\n",
    "Note that the data are already normalized.\n",
    "The time has been defined such that $t=0$ corresponds to the start of $G_{1}$. Assume that there is no $G_{0}$ state."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a32ef471",
   "metadata": {},
   "source": [
    "### Function definition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cfbb4aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def run_PCA(data, n_components=5, color_by='replicate', symbol_by='time', scale_data=True, offset=0.1, std_div=True):\n",
    "    \n",
    "    \"\"\"\n",
    "    Runs PCA on the provided data and generates a scatter plot of pairs of principal components.\n",
    "    It also returns a DataFrame containing the PCA loadings.\n",
    "\n",
    "    Parameters:\n",
    "    data: The gene expression data to analyze.\n",
    "    n_components: The number of principal components to calculate.\n",
    "    color_by: The metadata category to color the data points by.\n",
    "    symbol_by: The metadata category to symbolize the data points by.\n",
    "    scale_data: Whether to scale the data before PCA.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame containing the PCA loadings.\n",
    "    \"\"\"\n",
    "\n",
    "    data_tmp = data.copy()\n",
    "    data=np.log2(data+offset).to_numpy()\n",
    "    # Scale each sample of the data by mean and by standard deviation if specified\n",
    "    if scale_data:\n",
    "        data[:, :50]=data[:, :50]-data[:, :50].mean(1).reshape(-1, 1)\n",
    "        data[:, 50:]=data[:, 50:]-data[:, 50:].mean(1).reshape(-1, 1)\n",
    "        if std_div:\n",
    "            data[:, :50]=data[:, :50]/data[:, :50].std(1).reshape(-1, 1)\n",
    "            data[:, 50:]=data[:, 50:]/data[:, 50:].std(1).reshape(-1, 1)\n",
    "    else:\n",
    "        data = data\n",
    "\n",
    "        \n",
    "    # Fit the PCA model\n",
    "    pca = PCA(n_components=n_components)\n",
    "    pca.fit(data.T)\n",
    "\n",
    "    # Transform the data\n",
    "    pca_transformed = pca.transform(data.T)\n",
    "\n",
    "    # Create a DataFrame for the PCA results\n",
    "    labels = [f'PC{i+1}' for i in range(n_components)]\n",
    "    pca_df = pd.DataFrame(pca_transformed, columns=labels)\n",
    "    pca_df[color_by] = data_tmp.columns.get_level_values(color_by)\n",
    "    pca_df[symbol_by] = data_tmp.columns.get_level_values(symbol_by)\n",
    "\n",
    "    if color_by==\"time\":\n",
    "        pca_df[color_by]=np.round(pd.to_numeric(data_tmp.columns.get_level_values(color_by))/50 *20, 1).astype(str)\n",
    "        palette='twilight'\n",
    "    else:\n",
    "        palette='Set1'\n",
    "\n",
    "    # Plotting each pair of consecutive principal components\n",
    "    fig, axes = plt.subplots(1, n_components -1, figsize=(30, 15))\n",
    "    for i in range(n_components-1):\n",
    "        sns.scatterplot(ax=axes[i], data=pca_df, x=f'PC{i+1}', y=f'PC{i+2}', \n",
    "        hue=color_by, style=symbol_by, s=200, palette=palette)\n",
    "\n",
    "        axes[i].set_xlabel(f'PC{i+1} ({pca.explained_variance_ratio_[i]*100:.2f}%)')\n",
    "        axes[i].set_ylabel(f'PC{i+2} ({pca.explained_variance_ratio_[i+1]*100:.2f}%)')\n",
    "        axes[i].set_box_aspect(1) \n",
    "        # remove legend if it's not the first plot\n",
    "        if i == 0:\n",
    "        # Move legend outside for the first plot\n",
    "            axes[i].legend(loc='upper center', bbox_to_anchor=(1.3, 1), ncol=2, title=color_by) #\n",
    "        else:\n",
    "        # Remove legend for all other plots\n",
    "            axes[i].get_legend().remove()\n",
    "    fig.suptitle(f'PCA of Gene Expression Data (color={color_by}, symbol={symbol_by})', fontsize=16)\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    # Return the PCA loadings\n",
    "    df_pca = pd.DataFrame(pca.components_, columns=data_tmp.index, index=labels)\n",
    "\n",
    "    return df_pca\n",
    "\n",
    "def harm_reg(x, t, period):\n",
    "    \"\"\"\n",
    "    Fit a harmonic regression model to a time series. Perform likelihood ratio test between the full and null model. Get p-value from the test.\n",
    "    \n",
    "    Parameters:\n",
    "    x (array): Time series data\n",
    "    t (array): Time points\n",
    "    period (float): Period of the harmonic function\n",
    "\n",
    "    Returns:\n",
    "    tuple: Mean (mu), peak-to-through amplitude (amp), phase (phase), and p-value (p_val)\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    n = len(x)\n",
    "    \n",
    "    # Fit the null model (intercept only)\n",
    "    X0 = np.ones((n, 1))  # Design matrix for the null model\n",
    "    fit0 = sm.OLS(x, X0).fit()\n",
    "    \n",
    "    c = np.cos(2 * np.pi * t / period)\n",
    "    s = np.sin(2 * np.pi * t / period)\n",
    "    \n",
    "    # Design matrix for the full model\n",
    "    X1 = np.column_stack((np.ones(n), c, s))\n",
    "    fit1 = sm.OLS(x, X1).fit()\n",
    "    \n",
    "    # Coefficients\n",
    "    mu = fit1.params[0]\n",
    "    a = fit1.params[1]\n",
    "    b = fit1.params[2]\n",
    "    \n",
    "    # Likelihood ratio test\n",
    "    lr_stat = 2 * (fit1.llf - fit0.llf)\n",
    "    p_val = scipy.stats.chi2.sf(lr_stat, 2)\n",
    "    \n",
    "    # Amplitude and phase\n",
    "    amp = 2 * np.sqrt(a**2 + b**2)\n",
    "    phase = np.arctan2(b, a) % (2 * np.pi)\n",
    "    phase = period * phase / (2 * np.pi)\n",
    "\n",
    "    \n",
    "    return mu, amp, phase, p_val\n",
    "\n",
    "def fit_harmo(data, time, period):\n",
    "    \"\"\"\n",
    "    Applies the harm_reg function to all genes for a given normalized gene expression dataset.\n",
    "\n",
    "    Parameters:\n",
    "    data (pd.DataFrame): Normalized gene expression dataset with genes as rows.\n",
    "    time (array): Time points for the dataset.\n",
    "    period (float): Period of the harmonic function.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame containing the amplitude, phase, p-value, qvalue and mu for each gene.\n",
    "    \"\"\"\n",
    "    results = data.apply(lambda x: harm_reg(x.values, time, period), axis=1)\n",
    "    results_df = pd.DataFrame(results.tolist(), index=data.index, columns=['mu', 'amp', 'phase', 'p_val'])\n",
    "    results_df['p_val'] = results_df['p_val'].fillna(1)\n",
    "    results_df['q_val'] = multipletests(results_df['p_val'], method='fdr_bh')[1]\n",
    "\n",
    "    return results_df\n",
    "\n",
    "\n",
    "# Plotting function for side-by-side intron and exon plots\n",
    "def plot_gene_expression(dat_melted, gene_name):\n",
    "    \"\"\"\n",
    "    Plots the expression of a gene introns and exons over time.\n",
    "\n",
    "    Parameters:\n",
    "    dat_melted (pd.DataFrame): The melted gene expression dataset.\n",
    "    gene_name (str): The gene to plot.\n",
    "\n",
    "    Returns:\n",
    "    None\n",
    "    \"\"\"\n",
    "\n",
    "    # Filter the melted DataFrame for the gene of interest\n",
    "    data = dat_melted[dat_melted['Gene_Symbol'] == gene_name]\n",
    "\n",
    "    fig, axes = plt.subplots(1, 2, figsize=(12, 6), sharey=True)  # Create 1 row, 2 columns\n",
    "\n",
    "    # Plot Replicate 0 Intron and Exon Expression\n",
    "    sns.scatterplot(\n",
    "        data=data[data['replicate'] == \"0\"],  # Filter directly for intron\n",
    "        x='time', y='expression', hue='feature', style='feature',\n",
    "        alpha=0.6, legend='full', ax=axes[0]\n",
    "    )\n",
    "    sns.lineplot(\n",
    "        data=data[data['replicate'] == \"0\"],  # Filter directly for intron\n",
    "        x='time', y='expression', hue='feature', estimator='mean',\n",
    "        errorbar=None, legend=False, ax=axes[0]\n",
    "    )\n",
    "    axes[0].set_title(f\"{gene_name} Introns and Exons Expression\\n Over Time For Replicate 0\")\n",
    "    axes[0].set_xlabel(\"Time (hours)\")\n",
    "    axes[0].set_ylabel(\"Expression Level\")\n",
    "    axes[0].grid(alpha=0.3)\n",
    "    axes[0].set_box_aspect(1)  # Ensure square aspect ratio\n",
    "    axes[0].set_xticks(np.arange(0, 49, 5), np.linspace(0, 20, 11).astype(int)[:-1])\n",
    "\n",
    "    # Plot Replicate 1 Intron and Exon Expression\n",
    "    sns.scatterplot(\n",
    "        data=data[data['replicate'] == \"1\"],  # Filter directly for intron\n",
    "        x='time', y='expression', hue='feature', style='feature',\n",
    "        alpha=0.6, legend='full', ax=axes[1]\n",
    "    )\n",
    "    sns.lineplot(\n",
    "        data=data[data['replicate'] == \"1\"],  # Filter directly for intron\n",
    "        x='time', y='expression', hue='feature',\n",
    "        errorbar=None, legend=False, ax=axes[1]\n",
    "    )\n",
    "    axes[1].set_title(f\"{gene_name} Introns and Exons Expression\\n Over Time For Replicate 1\")\n",
    "    axes[1].set_xlabel(\"Time (hours)\")\n",
    "    axes[1].set_ylabel(\"Expression Level\")\n",
    "    axes[1].grid(alpha=0.3)\n",
    "    axes[1].set_box_aspect(1)  # Ensure square aspect ratio\n",
    "    axes[1].set_xticks(np.arange(0, 49, 5), np.linspace(0, 20, 11).astype(int)[:-1])\n",
    "\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "def best_subplot_shape(n_Genes):\n",
    "    n_plots=n_Genes\n",
    "    \"\"\"\n",
    "    Return (n_rows, n_cols) giving a near-square layout for n_plots subplots.\n",
    "    \"\"\"\n",
    "    n_cols = math.ceil(math.sqrt(n_plots))\n",
    "    n_rows = math.ceil(n_plots / n_cols)\n",
    "    return n_rows, n_cols\n",
    "def plot_gene_list(dat, all_genes_to_plot):\n",
    "    \"\"\"\n",
    "    Plots the levels of the introns and exons of the genes in your list over time.\n",
    "    \"\"\"\n",
    "    exon_0_cols=[c for c in dat.columns.tolist() if (\"exon\" in c[0] and c[3].split(\"_\")[-1]==\"0\")]\n",
    "    intron_0_cols=[c for c in dat.columns.tolist() if (\"intron\" in c[0] and c[3].split(\"_\")[-1]==\"0\")]\n",
    "    exon_1_cols=[c for c in dat.columns.tolist() if (\"exon\" in c[0] and c[3].split(\"_\")[-1]==\"1\")]\n",
    "    intron_1_cols=[c for c in dat.columns.tolist() if (\"intron\" in c[0] and c[3].split(\"_\")[-1]==\"1\")]\n",
    "    n_rows, n_cols=best_subplot_shape(len(all_genes_to_plot))\n",
    "    #plt.ioff()\n",
    "    fig, axs = plt.subplots(n_rows, n_cols, figsize=(n_cols*5, n_rows*5))  # 9 plots (PC1–PC2, ..., PC8–PC9)\n",
    "    axs = axs.flatten()\n",
    "    for g_ind, gene in enumerate(all_genes_to_plot):\n",
    "        ax=axs[g_ind]\n",
    "        ax_gs = ax.get_subplotspec().subgridspec(2, 1, hspace=0.175)\n",
    "        plt.delaxes(ax)  # remove the placeholder axes\n",
    "        ax_ex = fig.add_subplot(ax_gs[1])\n",
    "        ax_in = fig.add_subplot(ax_gs[0], sharex=ax_ex)\n",
    "        ax_in.plot(dat.loc[gene, intron_0_cols].to_numpy(), c=\"orange\", label=\"Introns_rep0\")\n",
    "        #ax_ex=ax.twinx()\n",
    "        ax_ex.plot(dat.loc[gene, exon_0_cols].to_numpy(), c=\"red\", label=\"Exons_rep0\")\n",
    "        ax_in.plot(dat.loc[gene, intron_1_cols].to_numpy(), c=\"lightblue\", label=\"Introns_rep1\")\n",
    "        ax_ex.plot(dat.loc[gene, exon_1_cols].to_numpy(), c=\"blue\", label=\"Exons_rep1\")\n",
    "        ax_in.set_title(gene)\n",
    "        ax_ex.set_xlabel(\"Time [Hr]\")\n",
    "        ax_ex.set_ylabel(\"Mean counts\")\n",
    "        ax_in.set_ylabel(\"Mean counts\")\n",
    "        ax_in.set_xticks(np.arange(0, 49, 5), np.linspace(0, 20, 11).astype(int)[:-1])\n",
    "        ax_ex.set_xticks(np.arange(0, 49, 5), np.linspace(0, 20, 11).astype(int)[:-1])\n",
    "        ax_in.legend()\n",
    "        ax_ex.legend()\n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ff2e8ad",
   "metadata": {},
   "source": [
    "### Question 1\n",
    "\n",
    "1. Using the biological knowledge about cell cycle, design a list of 15-20 genes whose mRNA expression profiles should cover all the different phases of the cell cycle.  \n",
    "For each gene, plot the levels through the cell cycle of both replicates for introns and exons.  \n",
    "Make sure you get at least one gene peaking in each phase of the cell cycle.  \n",
    "\n",
    "2. Perform PCA using your list of genes.  \n",
    "Start by analyzing only the exons, then focus on the introns.\n",
    "Comment on your observations. \n",
    "\n",
    "3. Using these plots (levels of each gene and PCA), provide estimates of when happen $G_{1}/S$ and $G_{2}/M$ transitions. \n",
    "\n",
    "4. Use the *fit_harmo* (i.e. harmonic regression) function on the data to assess and characterize genome-wide gene expression periodicity. Perform this analysis separately for the exons and introns.\n",
    "\n",
    "- Select periodic genes based on q-value thresholds (e.g., $q < 0.05$) and amplitude thresholds (e.g., $log_2$ peak-to-trough > 0.5 for introns and $log_2$ peak-to-trough > 0.25 for exons) for further analysis.\n",
    "- Provide histograms of the peak times (i.e. phase) and amplitudes ($log_2$ peak-to-trough) for the selected genes.\n",
    "- Examine whether you observe temporal patterns in gene expression.  \n",
    "\n",
    "5. What types of genes can you find in the list of periodic genes? Do you recover the genes of the list you designed in 1. ? \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d725b9ef",
   "metadata": {},
   "source": [
    "#### Load the RNA-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a51a74b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dat = pd.read_csv('./pseudobulk.txt', sep='\\t', index_col=0)\n",
    "dat.columns = dat.columns.str.split('_', expand = True) # split columns by '_'\n",
    "dat.columns.names = ['feature','measure','time','replicate'] # name the columns, multi-index\n",
    "dat=dat.iloc[:, (dat.columns.get_level_values('measure') == \"mean\")]\n",
    "dat.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac79e73f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#melting the data for plotting later\n",
    "dat_melted = dat.stack(level=[0, 1, 2, 3], future_stack=True).reset_index()\n",
    "dat_melted.columns = ['Gene_Symbol', 'feature', 'measure', 'time', 'replicate', 'expression']\n",
    "dat_melted.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6a8e739a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Subset the data for intron and exon\n",
    "idx = pd.IndexSlice\n",
    "dat_intron = dat.loc[:, idx['intron', 'mean', :, :]]\n",
    "dat_exon = dat.loc[:, idx['exon', 'mean', :, :]]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4adc336",
   "metadata": {},
   "source": [
    "#### Plot cell cycle related genes and perform the PCA on them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef3573cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Your currated list of gene\n",
    "#your_genes_list=[...]\n",
    "\n",
    "#Plot the genes in your list\n",
    "#plot_gene_list(...)\n",
    "\n",
    "#run_PCA ....\n",
    "#...\n",
    "#...\n",
    "#..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18464198",
   "metadata": {},
   "source": [
    "#### Assess gene expression periodicity genome-wide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "477fd66e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply the function to both intron and exon data.\n",
    "offset=0.1\n",
    "dat_intron_fit=dat_intron\n",
    "dat_exon_fit=dat_exon\n",
    "dat_intron_fit = np.log2(dat_intron_fit + offset)\n",
    "dat_exon_fit = np.log2(dat_exon_fit + offset)\n",
    "\n",
    "T = 20\n",
    "time = dat_exon_fit.columns.get_level_values('time').str.replace('ZT','').astype(int).to_numpy()\n",
    "time=time/(time.max()+1)\n",
    "time=time*T\n",
    "\n",
    "results_intron = fit_harmo(dat_intron_fit, time, T)\n",
    "results_exon = fit_harmo(dat_exon_fit, time, T)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bafa827",
   "metadata": {},
   "source": [
    "#### Define the genes of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18e0ffa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the genes with significant periodic gene expression in introns and exons\n",
    "#genes_per_introns = np.logical_and(..., ...)\n",
    "#...\n",
    "#..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b32da483",
   "metadata": {},
   "source": [
    "#### Plot phase and amplitude distribution for the rhythmic genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82d3579a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#...\n",
    "#...\n",
    "#...\n",
    "#..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b91055f1",
   "metadata": {},
   "source": [
    "#### Investigate the types of genes that were selected as periodic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ced63b90",
   "metadata": {},
   "outputs": [],
   "source": [
    "#...\n",
    "#...\n",
    "#...\n",
    "#..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cbe75a3",
   "metadata": {},
   "source": [
    "### Question 2\n",
    "\n",
    "1. Related to Exercise 1, study the relationship between the peak times (i.e., phase) of pre-mRNA and mRNA for genes that exhibit periodic behaviour at both the pre-mRNA and mRNA levels. Use appropriate cutoffs in your analysis to select a subset of genes that are periodic at both the pre-mRNA and mRNA levels. Aim to include several hundred genes for a robust analysis.  \n",
    "   **Hint:** Show a scatter plot of mRNA vs. pre-mRNA peak times. Explain your findings.\n",
    "\n",
    "2. Show representative profiles of genes with short or long delays between the pre-mRNA and mRNA peak times. Explain.\n",
    "\n",
    "3. From the delay between the pre-mRNA and mRNA ($\\delta t$ [hr]), infer the mRNA half-life ($\\tau$ [hr]) using the formula derived in the course:  \n",
    "\n",
    "\n",
    "   $$\\tan\\left(\\frac{2\\pi \\cdot \\delta t}{20}\\right) = \\frac{2\\pi}{20}  \\frac{ \\tau}{\\ln(2)}$$\n",
    " \n",
    "\n",
    "   Plot the histogram of the inferred mRNA half-lives and a scatter plot of half life vs mRNA amplitudes ($log_2$ peak-to-trough). Explain your findings.\n",
    "\n",
    "4. **(Bonus):** What can you infer about the splicing rate?  \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8e26736",
   "metadata": {},
   "source": [
    "#### Plot pre-mRNA, mRNA delay"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dff747a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot phase exons vs phase introns\n",
    "#...\n",
    "#...\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7292993",
   "metadata": {},
   "source": [
    "### Plot representative genes with short or long-delay"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4912a364",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Compute the delay between intron and exon\n",
    "#Restrict to genes with phase difference < 5 hours and with exon later than intron\n",
    "\n",
    "#...\n",
    "#...\n",
    "#...\n",
    "#...\n",
    "\n",
    "#find genes with short delay, plot representative gene expression\n",
    "#plot_gene_expression(dat_melted, \"GeneName\")\n",
    "\n",
    "\n",
    "#find gene with long delay between 3 and 5 hours\n",
    " #plot_gene_expression(dat_melted, \"GeneName\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a554df5",
   "metadata": {},
   "source": [
    "### Compute half-life from the delay between intron and exon\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a818efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#...\n",
    "#...\n",
    "#...\n",
    "#..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db956180",
   "metadata": {},
   "source": [
    "#### Compare Half lives and mRNA amplitudes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37ef6ce6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#...\n",
    "#...\n",
    "#...\n",
    "#..."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cellpose",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
